File: 3-leetcode/02-medium/cpp/885-spiral-matrix-iii.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart,
                                        int cStart)
    {
        vector<pair<int, int>> dir{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        vector<vector<int>> traversed;

        for (int step = 1, direction = 0; traversed.size() < rows * cols;)
        {
            for (int i = 0; i < 2; i++)
            {
                for (int j = 0; j < step; ++j)
                {
                    if (rStart >= 0 && rStart < rows && cStart >= 0 &&
                        cStart < cols)
                    {
                        traversed.push_back({rStart, cStart});
                    }
                    rStart += dir[direction].first;
                    cStart += dir[direction].second;
                }

                direction = (direction + 1) % 4;
            }
            step++;
        }
        return traversed;
    }
};

File: 3-leetcode/02-medium/js/885-spiral-matrix-iii.ts
const spiralMatrixIII = (
  rows: number,
  cols: number,
  rStart: number,
  cStart: number,
): number[][] => {
  const result = [];
  const dir = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0],
  ];

  for (let step = 1, direction = 0; result.length < rows * cols; ) {
    for (let i = 0; i < 2; i++) {
      for (let j = 0; j < step; j++) {
        if (rStart >= 0 && rStart < rows && cStart >= 0 && cStart < cols) {
          result.push([rStart, cStart]);
        }

        rStart += dir[direction][0];
        cStart += dir[direction][1];
      }

      direction = (direction + 1) % 4;
    }
    step++;
  }

  return result;
};
